src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+
+            if (_tokenType == JsonTokenType.None)
+                goto ReadFirstToken;
+
+            if (_tokenType == JsonTokenType.StartObject)
+            {
+                if (first == JsonConstants.CloseBrace)
+                    EndObject();
+                else
+                {
+                    if (first != JsonConstants.Quote)
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedStartOfPropertyNotFound, first);
+
+                    int prevConsumed = _consumed;
+                    long prevPosition = _lineBytePosition;
+                    long prevLineNumber = _lineNumber;
+                    retVal = ConsumePropertyName();
+                    if (!retVal)
+                    {
+                        // roll back potential changes
+                        _consumed = prevConsumed;
+                        _tokenType = JsonTokenType.StartObject;
+                        _lineBytePosition = prevPosition;
+                        _lineNumber = prevLineNumber;
+                    }
+                    goto Done;
+                }
+            }
+            else if (_tokenType == JsonTokenType.StartArray)
+            {
+                if (first == JsonConstants.CloseBracket)
+                    EndArray();
+                else
+                {
+                    retVal = ConsumeValue(first);
+                    goto Done;
+                }
+            }
+            else if (_tokenType == JsonTokenType.PropertyName)
+            {
+                retVal = ConsumeValue(first);
+                goto Done;
+            }
+            else
+            {
+                int prevConsumed = _consumed;
+                long prevPosition = _lineBytePosition;
+                long prevLineNumber = _lineNumber;
+                JsonTokenType prevTokenType = _tokenType;
+                ConsumeTokenResult result = ConsumeNextToken(first);
+                if (result == ConsumeTokenResult.Success)
+                {
+                    retVal = true;
+                    goto Done;
+                }
+                if (result == ConsumeTokenResult.IncompleteRollback)
+                {
+                    _consumed = prevConsumed;
+                    _tokenType = prevTokenType;
+                    _lineBytePosition = prevPosition;
+                    _lineNumber = prevLineNumber;
+                }
+                goto Done;
+            }
+
+            retVal = true;
+
+        Done:
+            return retVal;
+
+        ReadFirstToken:
+            retVal = ReadFirstToken(first);
+            goto Done;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool HasMoreData()
+        {
+            if (_consumed >= (uint)_buffer.Length)
+            {
+                if (_isNotPrimitive && IsLastSpan)
+                {
+                    if (_tokenType != JsonTokenType.EndArray && _tokenType != JsonTokenType.EndObject)
```

Comment:
Any reason for the nested if rather than just having all these conditions in the one if?

Reply:
Not really. I do have the `if (_consumed >= (uint)_buffer.Length)` pattern in several other places where the inner conditions are different so I don't necessarily want to merge all three, but I could merge the two nested ifs.